### PR TITLE
Introduce non-classpath resource upload mechanism

### PR DESCRIPTION
**Pull request by eminn**
Provides convenience for users to attach their files and directories in the Jet pipelines.  

Checklist
- [x] Tags Set
- [x] Milestone Set
- [x] Any breaking changes are documented
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc
- [ ] For code samples, code sample main readme is updated

Breaking Changes:
- Removed `REGULAR_FILE` type and added `CLASSPATH_RESOURCE` ,`FILE` and `DIRECTORY` to `ResourceType` enum 
- Renamed `JobConfig.addResource` to `addClasspathResource`

Co-authored-by: Marko Topolnik <marko@hazelcast.com>

### CODE REVIEW

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java [hunk 1]
```diff
@@ -393,5 +407,35 @@ default boolean snapshottingEnabled() {
...
+        /**
+         * Returns a temporary directory which contains the attached files to the
+         * job with the given id.
+         *
+         * @param id identifier defined on the {@link ResourceConfig} to be
+         *           used retrieve files from storage.
+         * @return {@link File} handle to a temporary directory which contains the
+         * files attached to the job with provided identifier.
+         */
+        default File getJobFileStorageById(String id) {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java]: shouldn't this be part of the context ? It seems strange to add it to PMS? Is it meant to be overriden?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ResourceConfig.java [hunk 2]
```diff
@@ -40,8 +41,8 @@
      * @param id            id of the resource
      * @param resourceType  type of the resource
      */
-    ResourceConfig(@Nonnull URL url, String id, ResourceType resourceType) {
-        checkTrue((resourceType != ResourceType.REGULAR_FILE) ^ id != null,
+    ResourceConfig(@Nonnull URL url, @Nullable String id, @Nonnull ResourceType resourceType) {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/ResourceConfig.java]: would it be better to always assign id here?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java [hunk 3]
```diff
@@ -282,7 +284,8 @@ public void init(@Nonnull Context context) {
                                       .getPartition(key).getOwner().getAddress();
             }
 
-            @Nonnull @Override
+            @Nonnull
+            @Override
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java]: this change can be reverted

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java [hunk 4]
```diff
@@ -393,5 +396,28 @@ default boolean snapshottingEnabled() {
...
+        /**
+         * Returns the attached directory to the job with the given id.
+         *
+         * @param id identifier defined on the {@link ResourceConfig} to be
+         *           used retrieve files from storage.
+         * @return {@link File} handle to the attached directory to the job with provided identifier.
+         * @since 4.0
+         */
+        @Nonnull
+        File getAttachedDirectory(@Nonnull String id);
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/ProcessorMetaSupplier.java]: you can skip the `get` (so just have `attachedDirectory` as it's the convention on this interface

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java [hunk 5]
```diff
@@ -351,6 +373,19 @@ void deleteJob(long jobId) {
         // delete the job record and related records
         jobExecutionRecords.remove(jobId);
         jobRecords.remove(jobId);
+        deleteTempFiles("jet-" + instance.getName() + "-" + idToString(jobId));
+    }
+
+    private void deleteTempFiles(String filter) {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java]: This seems rather expensive - it's going to iterate all temp files each time a job completes? We made a lot of optimizations for jobs that take just a few ms (small query jobs). It could be done as part of the Job GC process. Furthermore this is only called on master node, not on other nodes, how are those files on other nodes cleaned up?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java [hunk 6]
```diff
@@ -182,15 +192,25 @@ long uploadJobResources(JobConfig jobConfig) {
...
+                            readStreamAndPutCompressedToMap(rc.getId(), tmpMap, in);
+                        }
+                        break;
+                    case FILE:
+                        IMapOutputStream os = new IMapOutputStream(getJobFileStorage(jobId).get(),
+                                idToString(jobId) + rc.getId());
+                        zipFileToOutputStream(Paths.get(rc.getUrl().getFile()), os);
+                        break;
+                    case DIRECTORY:
+                        IMapOutputStream os2 = new IMapOutputStream(getJobFileStorage(jobId).get(),
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java]: is this stream closed somewhere?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java [hunk 7]
```diff
@@ -246,6 +257,83 @@ public static void checkSerializable(Object object, String objectName) {
...
+        try (ZipOutputStream zipOut = new ZipOutputStream(outputStream)) {
+            Path fileName = filePath.getFileName();
+            if (fileName != null) {
+                zipOut.putNextEntry(new ZipEntry(fileName.toString()));
+                try (InputStream in = Files.newInputStream(filePath)) {
+                    copyStream(in, zipOut);
+                }
+                zipOut.closeEntry();
+            }
+        } catch (IOException e) {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java]: you can make the methods throws IOException , as it's already handled above.

**Comment by cangencer** [comment]
We need unit tests for `IMapOutputStream`, testing edge cases.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java [hunk 8]
```diff
@@ -418,6 +558,12 @@ public JobConfig addResource(@Nonnull String path, @Nonnull String id) {
     }
 
     private JobConfig add(URL url, String id, ResourceType resourceType) {
+        boolean duplicate = resourceConfigs.stream()
+                                           .map(ResourceConfig::getId)
+                                           .anyMatch(existingId -> existingId.equals(id));
+        if (duplicate) {
+            throw new IllegalArgumentException("Resource with id:" + id + " already exists!");
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java]: can we remove the `!` marks from exceptions? There's several in many places in the PR

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java [hunk 9]
```diff
@@ -418,6 +558,12 @@ public JobConfig addResource(@Nonnull String path, @Nonnull String id) {
     }
 
     private JobConfig add(URL url, String id, ResourceType resourceType) {
+        boolean duplicate = resourceConfigs.stream()
+                                           .map(ResourceConfig::getId)
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java]: can be simplified like `.anyMatch(c -> id.equals(c.getId())`

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java [hunk 10]
```diff
@@ -108,26 +122,69 @@ public ILogger logger() {
...
         }
+
+        @Nonnull @Override
+        public File attachedDirectory(@Nonnull String id) {
+            Preconditions.checkHasText(id, "id cannot be null or empty!");
+            findResourceConfigOrThrowException(id);
+            JetInstance instance = jetInstance();
+            String jobId = idToString(jobId());
+            IMap<String, byte[]> map = instance.getMap(FILE_STORAGE_MAP_NAME_PREFIX + jobId);
+            try (IMapInputStream inputStream = new IMapInputStream(map, jobId, id)) {
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java]: wouldn't this create the file everytime it's asked?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ExecutionContext.java [hunk 11]
```diff
@@ -63,6 +66,7 @@
     private final Object executionLock = new Object();
     private final ILogger logger;
     private String jobName;
+    private List<File> localFiles = new ArrayList<>();
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ExecutionContext.java]: this list is accessed concurrently, but is not thread-safe

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java [hunk 12]
```diff
@@ -108,26 +122,69 @@ public ILogger logger() {
         public ProcessingGuarantee processingGuarantee() {
             return processingGuarantee;
         }
-    }
+
+           }
```
Comment (cangencer) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java]: wrong indentation

**Comment by viliam-durina** [comment]
Is any of the listed breaking changes actually a breaking change? They are either an addition or are in private API.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java [hunk 13]
```diff
@@ -417,7 +570,14 @@ public JobConfig addResource(@Nonnull String path, @Nonnull String id) {
         return resourceConfigs;
     }
 
-    private JobConfig add(URL url, String id, ResourceType resourceType) {
+    private JobConfig add(@Nonnull URL url, @Nullable String id, @Nonnull ResourceType resourceType) {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/config/JobConfig.java]: the `id` should be `@Nonnull` too, I checked all callers and they ensure it. Even if it's null, we pass it to `ResourceConfig` constructor where it's non-null.

**Comment by cangencer** [comment]
@viliam-durina `JobConfig.addResource` was not private. But yes the description mixes breaking changes with non-breaking ones.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/test/TestProcessorSupplierContext.java [hunk 14]
```diff
@@ -71,9 +75,24 @@ public int memberIndex() {
         return memberIndex;
     }
 
-    /**
-     * Set the member index.
-     */
+    @Nonnull @Override
+    public File attachedDirectory(@Nonnull String id) {
+        return attachedDirectory;
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/test/TestProcessorSupplierContext.java]: We should store the files in a map, keyed by the `id`. This might be enough fur current cases, but it's an incorrect implementation.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/deployment/IMapOutputStream.java [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Override
+    public void flush() throws IOException {
+        if (isClosed()) {
+            return;
+        }
+        boolean incompleteChunk = currentChunk.remaining() > 0;
+        byte[] value = currentChunk.array();
+        if (incompleteChunk) {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/deployment/IMapOutputStream.java]: Why this test? If there's no incomplete chunk, we'll store the entire backing array into the imap but we won't increment the chunk index and we'll overwrite it next time (if there is a _next time_)

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/deployment/IMapOutputStream.java [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+
+public class IMapOutputStream extends OutputStream {
+    private static final int CHUNK_SIZE = 1 << 17;
+
+    private final String prefix;
+    private final IMap<String, byte[]> map;
+    private final ByteBuffer currentChunk = ByteBuffer.allocate(CHUNK_SIZE);
+    private final byte[] singleByteBuffer = new byte[1];
+
+    private int currentChunkIndex = 1;
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/deployment/IMapOutputStream.java]: If you started from 0, you'll have less plus-ones and minus-ones in the code.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java [hunk 17]
```diff
@@ -182,15 +192,23 @@ long uploadJobResources(JobConfig jobConfig) {
...
-                        readStreamAndPutCompressedToMap(rc.getId(), tmpMap, in);
-                        in.close();
+                    case CLASS:
+                        try (InputStream in = rc.getUrl().openStream()) {
+                            readStreamAndPutCompressedToMap(rc.getId(), tmpMap, in);
+                        }
+                        break;
+                    case FILE:
+                        IMapOutputStream os = new IMapOutputStream(getJobFileStorage(jobId).get(), rc.getId());
+                        zipFileToOutputStream(Paths.get(rc.getUrl().getFile()), os);
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java]: What if the URL isn't a local file? It will break. We should use `rc.getUrl().openStream()`. 

Also the `zipFileToOutputStream` is named incorrectly. What it does is that it puts a file as a sole entry in a zip file.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java [hunk 18]
```diff
@@ -108,26 +131,78 @@ public ILogger logger() {
...
         }
 
         @Override
         public int memberIndex() {
             return memberIndex;
         }
+
+        @Nonnull @Override
+        public File attachedDirectory(@Nonnull String id) {
+            Preconditions.checkHasText(id, "id cannot be null or empty");
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java]: We don't require non-empty ID when adding files. Also again 7 lines below.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java [hunk 19]
```diff
@@ -108,26 +131,78 @@ public ILogger logger() {
...
+            findResourceConfigOrThrowException(id);
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java]: Why don't we just rely on the presence of the key in the file IMap? This is superfluous. I was even thinking whether we shouldn't make the `resourceConfigs` transient in `JobConfig`, the URLs might be invalid on another member. The ID is enough and the user is supposed to know the IDs, we don't need to have them on members.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java [hunk 20]
```diff
@@ -246,6 +257,89 @@ public static void checkSerializable(Object object, String objectName) {
...
+        }
+    }
+
+    /**
+     * Creates a ZIP-file stream from the directory tree rooted at the supplied
+     * {@code baseDir}. Pipes the stream into the provided output stream, closing
+     * it when done.
+     * <p>
+     * <strong>Note:</strong> files and directories starting with either {@code _}
+     * or {@code .} are ignored.
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java]: - This should be documented at `JobConfig.attachDirectory`, not here. 
- It is a non-portable way to detect hidden files, it's valid on macOS only i guess (never heard about the `_` prefix). You should use `Files.isHidden(path)`.
- You should also ignore hidden files, not just hidden directories.
- It might be easier to use `Files.find`

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java [hunk 21]
```diff
@@ -182,15 +192,23 @@ long uploadJobResources(JobConfig jobConfig) {
...
-                        InputStream in = rc.getUrl().openStream();
-                        readStreamAndPutCompressedToMap(rc.getId(), tmpMap, in);
-                        in.close();
+                    case CLASS:
+                        try (InputStream in = rc.getUrl().openStream()) {
+                            readStreamAndPutCompressedToMap(rc.getId(), tmpMap, in);
+                        }
+                        break;
+                    case FILE:
+                        IMapOutputStream os = new IMapOutputStream(getJobFileStorage(jobId).get(), rc.getId());
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java]: This is broken, we get a lazy concurrent supplier each time and immediately do a `get()` on it. The `getJobFileStorage(jobId)` should be stored outside of the loop. And it doesn't have to be concurrent.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java [hunk 22]
```diff
@@ -458,6 +481,13 @@ public JobExecutionRecord getJobExecutionRecord(long jobId) {
         return Util.memoizeConcurrent(() -> instance.getMap(RESOURCES_MAP_NAME_PREFIX + idToString(jobId)));
     }
 
+    /**
+     * Gets the job files storage map, lazily evaluated to avoid creating the map if it won't be needed
+     */
+    Supplier<IMap<String, byte[]>> getJobFileStorage(long jobId) {
+        return Util.memoizeConcurrent(() -> instance.getMap(jobFileStorageMapName(jobId)));
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java]: We should make it lazy at caller, not here. Here we can get the map unconditionally.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/deployment/JetClassLoader.java [hunk 23]
```diff
@@ -76,28 +66,6 @@ public JetClassLoader(@Nonnull NodeEngine nodeEngine,
         return defineClass(name, classBytes, 0, classBytes.length);
     }
 
-    @Override
-    protected URL findResource(String name) {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/deployment/JetClassLoader.java]: Why was it necessary to remove this method? What if there's a jar on classpath and it wants to read some xml file on its classpath?

File: hazelcast-jet-core/src/test/java/com/hazelcast/jet/config/ResourceConfigTest.java [hunk 24]
```diff
@@ -0,0 +1,327 @@
...
+        config.attachFile(path, resourceName);
+        ResourceConfig resourceConfig = config.getResourceConfigs().iterator().next();
+
+        assertEquals(resourceName, resourceConfig.getId());
+        assertEquals(ResourceType.FILE, resourceConfig.getResourceType());
+        assertEquals(new File(path).toURI().toURL(), resourceConfig.getUrl());
+    }
+
+    @Test
+    public void testAttachFile_with_File() throws Exception {
```
Comment (viliam-durina) [inline hazelcast-jet-core/src/test/java/com/hazelcast/jet/config/ResourceConfigTest.java]: This test is exactly the same as `JobConfigTest.when_addResourceWithFile_thenReturnsResourceConfig`. Others probably too.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java (hunk 17 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java]: It's worse, it actually fails on Windows with:
```
java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/work/hazelcast/hazelcast-jet/hazelcast-jet-core/target/test-classes/deployment/resource.txt

	at java.base/sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
	at java.base/sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
	at java.base/sun.nio.fs.WindowsPath.parse(WindowsPath.java:92)
	at java.base/sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:229)
	at java.base/java.nio.file.Path.of(Path.java:147)
	at java.base/java.nio.file.Paths.get(Paths.java:69)
	at com.hazelcast.jet.impl.JobRepository.uploadJobResources(JobRepository.java:202)
	at com.hazelcast.jet.impl.AbstractJetInstance.uploadResourcesAndAssignId(AbstractJetInstance.java:149)
	at com.hazelcast.jet.impl.AbstractJetInstance.newJob(AbstractJetInstance.java:57)
```
We must use `URL.openStream()`.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java (hunk 20 above)
Comment (cangencer) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/util/Util.java]: I've partially fixed this. Javadoc still needs updating.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java [hunk 25]
```diff
@@ -97,6 +102,11 @@
      */
     public static final String RESOURCES_MAP_NAME_PREFIX = INTERNAL_JET_OBJECTS_PREFIX + "resources.";
 
+    /**
+     * Name of internal IMap which stores attached job files.
+     */
+    public static final String FILE_STORAGE_MAP_NAME_PREFIX = INTERNAL_JET_OBJECTS_PREFIX + "files.";
```
[the remark this answers is not in the document]
Comment (cangencer) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java]: we can do this using a namespace as part of the key (i.e. class.<id> or file.<id>) so there's no interference between CL and file resources. WDYT?

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/deployment/IMapOutputStream.java (hunk 15 above)
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/deployment/IMapOutputStream.java]: Got it, I was wrong. It writes an entry and then overwrites that entry with more data. Hopefully, nobody flushes mid-way of writing. But if somebody flushes just before closing, the final entry will be `put` twice.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java [hunk 26]
```diff
@@ -108,26 +131,76 @@ public ILogger logger() {
...
 
         @Override
         public int memberIndex() {
             return memberIndex;
         }
+
+        @Nonnull @Override
+        public File attachedDirectory(@Nonnull String id) {
+            Preconditions.checkHasText(id, "id cannot be null or empty");
+            return tempDirectories.computeIfAbsent(id, x -> extractFileToDisk(id));
```
Comment (eminn) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java]: `computIfAbsent` method doesn't guarantee that the value object generated by the mapping function will end up as a value in the map. It uses `putIfAbsent` internally and if that fails it fetches the actual value from the map. The value object created with the mapping function will be left on the disk untracked.

Comment (cangencer) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java]: That's not true, CHM has a special implementation of `computeIfAbsent` which is atomic.

Comment (eminn) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java]: I've missed that, I was looking at the default implementation on the ConcurrentMap. Good to know that.

Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/init/Contexts.java]: We should then use CHM for the field, not just CM.
